brave/src/main/java/brave/sampler/RateLimitingSampler.java

```diff
@@ -0,0 +1,43 @@
+package brave.sampler;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Based on https://github.com/aws/aws-xray-sdk-java/blob/2.0.1/aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/sampling/reservoir/Reservoir.java
+ *
+ * Uses only Java 6 APIs to be compatible with Brave Core requirements. Accounts for nanoTime()
+ * overflow, as unlikely as that might be.
+ */
+public class RateLimitingSampler extends Sampler {
+  public static Sampler create(int tracesPerSecond) {
```

Comment:
is it better use duration (pass long and TimeUnit), it will make the implementation become a bit complex but we have more friendly interface.

Reply:
I was considering that this morning. That logic could all be handled in the constructor so that the sample logic is the same. I'll give it a shot in the morning now that Adrian has provided benchmark scripts to check with

Reply:
I think we should add long and timeunit eventhough some timeunits are weird :) I have a different reason, though. You'll notice the base type Sampler has a create method. you can easily accidentally end up with a different impl if you use the wrong numeric type.

Changing the arity of the method would also solve that even if only units at or larger than second are allowed. One of the issues with having raw units though is that you can end up with fractional conversion if someone expresses something like 10/minute :P

IOTW I can see it both ways, I definitely think per second is easiest, and that is also the same as amazon accept fwiw

Reply:
@trustin @anuraaga any thoughts on this one?

Reply:
another thought is we use a builder.. one thing left dangling here is that the reset interval is default to second which is very easy impl as it is same unit as rate. reset interval of smaller can be helpful for people with high throughput, but again you end up with how to split the fractions if things are not aligned.

ex 35span/second reset interval of 10millis, you have to choose to have some buckets include zero and others one. This can be a separate change, as the logic would grow quite a bit.

Reply:
do we gonna make a builder?

Reply:
I think for now it is ok to use as-is. It prevents people from making units less than one second